google/cloud/internal/opentelemetry_test.cc

```diff
@@ -94,6 +99,57 @@ TEST(OpenTelemetry, MakeSpanWithAttributes) {
                          OTelAttribute<std::string>("key", "value"))));
 }
 
+TEST(OpenTelemetry, MakeSpanWithLink) {
+  auto span_catcher = InstallSpanCatcher();
+
+  auto s1 = MakeSpan("span1");
+  auto s2 = MakeSpan("span2", {}, {{s1->GetContext(), {{"key", "value"}}}});
+  s1->End();
+  s2->End();
+
+  auto spans = span_catcher->GetSpans();
+  EXPECT_THAT(
+      spans,
+      Contains(AllOf(
+          SpanNamed("span2"),
+          SpanHasLinks(AllOf(LinkHasSpanContext(s1->GetContext()),
+                             SpanLinkAttributesAre(OTelAttribute<std::string>(
+                                 "key", "value")))))));
+}
+
+TEST(OpenTelemetry, MakeSpanWithKind) {
+  auto span_catcher = InstallSpanCatcher();
+
+  auto s1 = MakeSpan("span1");
+  opentelemetry::trace::StartSpanOptions options;
+  options.kind = opentelemetry::trace::SpanKind::kProducer;
+  s1->End();
+
+  auto spans = span_catcher->GetSpans();
+  EXPECT_THAT(spans, Each(SpanKindIsClient()));
+}
+
+TEST(OpenTelemetry, MakeSpanWithParent) {
+  auto span_catcher = InstallSpanCatcher();
+
+  auto s1 = MakeSpan("span1");
+  opentelemetry::trace::StartSpanOptions options;
+  options.parent = s1->GetContext();
+  auto s2 = MakeSpan("span2", absl::make_optional(options));
+  s1->End();
+  s2->End();
+
+  auto spans = span_catcher->GetSpans();
+  ASSERT_EQ(2, spans.size());
+  // Span data will be in the order the spans are ended. The spans are
+  // returned in a `vector<unique_ptr<SpanData>>`, so we can't extract a span
+  // directly to get the span id from the span data. Instead we are accessing
+  // the span data using the `.at()`.
```

Comment:
Q. Why do we need `spans.at(0)` instead of `spans[0]`?

Reply:
No reason. Both are valid. Just saying we need to access it in the vector, and can not do something like

```
auto span_data1 = spans[0]
  EXPECT_THAT(spans,
              Contains(AllOf(SpanNamed("span2"),
                             SpanWithParentSpanId(span_data1->GetSpanId()))));
```

Reply:
OK.  We should just avoid saying `.at(i)` when there is no question about the validity of the index (and also question its use when throwing is itself questionable).

Reply:
Removed.